### PR TITLE
refactor: move validate to job validate

### PR DESCRIPTION
**Pull request by frrist**
## What
The initial intention of the PR was to simply move the `validate` command under `job` s.t. `job validate` became the new one. This new command is meant to validate a v2 job spec/models job spec. However it grew slightly in scope since we've never actually done real validation and I didn't want to just move a broken command.

The previous version of this command generated a job spec from the job model:
**Generate**:
https://github.com/bacalhau-project/bacalhau/blob/2e0c32949bba96bce07bcf50df14328ac1900463/cmd/cli/validate/validate.go#L165
**Job Model**
https://github.com/bacalhau-project/bacalhau/blob/2e0c32949bba96bce07bcf50df14328ac1900463/pkg/model/job.go#L191-L239

However since the previous job model didn't use the `json_schema` tags for generating a spec, and since all fields contained `omittempty` tags there wasn't really any validation happening since nearly all fields were incorrectly considered optional. e.g. in v1.3.2 `bacalhau validate` considers the below job file valid:
```yaml
 APIVersion: v1beta1
Spec:
  Engine: docker
  Deal:
    Concurrency: 1
```

## The Change
In this PR I have implemented a complete json schema describing our new job model which lives in `schemas/job/schemas`. This schema is then embedded in the bacalhau application when it runs. The motivation here is to have a `validate` command that actually performs validation. This schema aligns with the existing validation logic we have defined in go code. Ideally we will replace the manual go validation code with the requirements of this schema at some point in the future.

### Caveats
1. The schema requires all yaml/json keys be capitalized.
2. The schema requires (almost) all values (like `docker`, `wasm`, `local`, `s3`, `ipfs`, etc) align with the definitions in the code which are typically lower case.
3. This schema will need adjustment/further development if/when we introduce pluggable executors/storages/publishers since their schema is not represented.
4. This schema only validates users submitted job specs. Since the system representation of a job spec contains fields that users cannot set we do validate they are not present, but we cannot re-use this schema to validate a system representation of a job. Some prior discussion around system v submission job specs here: https://github.com/bacalhau-project/bacalhau/issues/3983

Since this is an improvement over the prior behavior this feels acceptable. And if users take issue with this we can do some per-processing on their behalf.

### CODE REVIEW

File: cmd/cli/job/validate.go
Comment (aronchick) [inline cmd/cli/job/validate.go:66]: the "job spec" is not valid...

File: pkg/lib/schemaloader/embedded.go
Comment (aronchick) [inline pkg/lib/schemaloader/embedded.go:48]: not that i care that much ... but why do we care? either it unmarshals or it doesn't

File: schemas/job/schemas/eng-docker.schema.json
Comment (aronchick) [inline schemas/job/schemas/eng-docker.schema.json:2]: aren't these generate-able via openapi?

**Review by aronchick — APPROVED** [review]
very useful - we should host this on a server that just returns these

**Review by wdbaruni — COMMENTED** [review]
Thanks for diving into this. There are few majors issues:

1. I think you've missed to include `job-schemas.json` file. 

2. The schema validation is enforcing the executors, publishers and storage sources we work with. We are no longer pluggable or modular where allow different networks to deploy their own engines. We've discussed this before as having this level of flexibility will limit the type of _client side_ validation we can do. There might be ways to download schemas at runtime from the requester, but thats a bigger topic and needs evaluating different user scenarios to see if it makes sense, as it will enforce the requester to know more about the engines, which it doesn't have to today.

3. Our schema validation is more than just checking field types and whether they exist or not. In `Task` validation for example we check that input source aliases are unique:
https://github.com/bacalhau-project/bacalhau/blob/4bea9885b7e8c3cca49b17cf901df3a947efed95/pkg/models/task.go#L143-L149

In job validation we should only validate if `Count` > 0 for `Batch` and `Service` jobs, and that it equals to zero for `Ops` and `Daemon` jobs. I don't think we do that today, but we should. 

These are achievable with json schema validations, but not trivial and not implemented in this version


4. We already have a [job.ValidateSubmission](https://github.com/bacalhau-project/bacalhau/blob/4bea9885b7e8c3cca49b17cf901df3a947efed95/pkg/models/job.go#L248) method that is used to validate jobs on the [client side ](https://github.com/bacalhau-project/bacalhau/blob/4bea9885b7e8c3cca49b17cf901df3a947efed95/cmd/cli/job/run.go#L144), [server side](https://github.com/bacalhau-project/bacalhau/blob/4bea9885b7e8c3cca49b17cf901df3a947efed95/pkg/publicapi/apimodels/job.go#L22), and also on the [backend side](https://github.com/bacalhau-project/bacalhau/blob/4bea9885b7e8c3cca49b17cf901df3a947efed95/pkg/models/job.go#L211) for additional validation before writing and reading from the datastore

Now we have different validation paths, unless the intention is to have those methods use schema validators as well. Then I might be concerned about the performance implications as we do call validate a few times. Also we usually call validate after the job is constructed and after calling `Normalize()`. 

5. Why couldn't you just use `job.ValidateSubmission` in `job validate` instead of implementing json validation now? The main value with json schema validation is that it is language agnostic and we can run the same validations on different SDKs for example, but they are more difficult to write compared to just validating in go, and I am not sure if we want to have this added complexity at this phase of the project.

6. Can you share examples of validation error messages to understand the user experience and how helpful these error messages are?

My overall feedback is to use the existing `job.ValidateSubmission` to unblock the API/CLI migration, and avoid adding major changes a week before the launch. We can have better discussion on the right thing to do when we have more room

**Comment by frrist** [comment]
> I think you've missed to include job-schemas.json file.

ooOooo great catch, thanks - this explains a lot (we had an overzealous ignore directive) https://github.com/bacalhau-project/bacalhau/blob/main/.gitignore#L49. I have addressed this in the latest commit and added a test to ensure the validate command continues to work.

> The schema validation is enforcing the executors, publishers and storage sources we work with.

Correct, the current schema validates the engines, storages, and publishers we currently support.

> We are no longer plugable or modular where allow different networks to deploy their own engines

We never were plugable in the first place. At present we don't support plugable engine, storage, or publisher types. In fact we return a 500 error if someone tries to use an engine that we don't support, for example ([job spec](https://github.com/bacalhau-project/bacalhau/blob/main/testdata/jobs/custom-task-type.yaml)):
```
> bacalhau job run testdata/jobs/custom-task-type.yaml 
Error: failed request: Unexpected response code: 500 ({
  "error": "failed to translate job type: flibble: unknown task type identified in translation: 'flibble'",
  "message": "Internal Server Error"
})
```

> We've discussed this before as having this level of flexibility will limit the type of client side validation we can do. There might be ways to download schemas at runtime from the requester, but thats a bigger topic and needs evaluating different user scenarios to see if it makes sense, as it will enforce the requester to know more about the engines, which it doesn't have to today.

Agreed, validation of plugable types is a broader topic. And since we don't support plugable types now I agree we should flesh this out at a later point.

> Our schema validation is more than just checking field types and whether they exist or not. In Task validation for example we check that input source aliases are unique.

Agreed, and this brings up a good point - or current validation is somewhat opaque to users. They must read through the go code to gather an understanding of what a valid job spec is. A benefit of adding schemas is that we have defined the rules of what's valid and what isn't in a canonical/standard manner.

> We already have a [job.ValidateSubmission](https://github.com/bacalhau-project/bacalhau/blob/4bea9885b7e8c3cca49b17cf901df3a947efed95/pkg/models/job.go#L248) method that is used to validate jobs on the [client side ](https://github.com/bacalhau-project/bacalhau/blob/4bea9885b7e8c3cca49b17cf901df3a947efed95/cmd/cli/job/run.go#L144), [server side](https://github.com/bacalhau-project/bacalhau/blob/4bea9885b7e8c3cca49b17cf901df3a947efed95/pkg/publicapi/apimodels/job.go#L22), and also on the [backend side](https://github.com/bacalhau-project/bacalhau/blob/4bea9885b7e8c3cca49b17cf901df3a947efed95/pkg/models/job.go#L211) for additional validation before writing and reading from the datastore

Indeed. And to be clear I think this method still adds value, and is something we should continue to use. But where it falls short is validating that unsupported fields  weren't included in the job spec, for example consider (`Kind`, `ThisIsntAllowd`, and `Selector`):
```
> bacalhau job run
Reading from /dev/stdin; send Ctrl-d to stop.Namespace: default
Count: 1
Type: batch
Kind: whatever
ThisIsntAllowed: but it is
Selector: fobar
Tasks:
  - Name: main
    Engine:
      Type: docker
      Params:
        Image: ubuntu:latest

Job successfully submitted. Job ID: j-b54ed6c2-548c-41aa-8cce-c4a135acf51a
Checking job status... (Enter Ctrl+C to exit at any time, your job will continue running):

        Communicating with the network  ................  done ✅  0.0s
           Creating job for submission  ................  done ✅  0.5s

To get more details about the run, execute:
        ./bin/linux/amd64/bacalhau job describe j-b54ed6c2-548c-41aa-8cce-c4a135acf51a

To get more details about the run executions, execute:
        ./bin/linux/amd64/bacalhau job executions j-b54ed6c2-548c-41aa-8cce-c4a135acf51a

```
Comparatively, when the same job is provided to `job validate` the user is given clear error messages:
```
> bacalhau job validate bad_job.yaml 
The Job is not valid. See errors:
- (root): Name is required
- (root): Additional property Kind is not allowed
- (root): Additional property Selector is not allowed
- (root): Additional property ThisIsntAllowed is not allowed
```  

> Why couldn't you just use job.ValidateSubmission in job validate instead of implementing json validation now

We can, and will make the change if you feel strongly about it. I didn't initially because the original validate command used a json schema for validation (as outlined in the PR description) and I intended to adopt a similar pattern for consistency.

> Can you share examples of validation error messages to understand the user experience and how helpful these error messages are?

Yup, see the above example.

> My overall feedback is to use the existing job.ValidateSubmission to unblock the API/CLI migration, and avoid adding major changes a week before the launch. We can have better discussion on the right thing to do when we have more room.

Is my understanding correct that you are blocking this PR until the schema is replaced with `job.ValidateSubmission`? Or is there something else that is blocking?

File: schemas/job/schemas/eng.schema.json
Comment (frrist) [reply schemas/job/schemas/eng.schema.json:14]: Yup, this is called out explicitly in the PR description and discussed in greater detail here: https://github.com/bacalhau-project/bacalhau/pull/4074#issuecomment-2168661112

File: schemas/job/schemas/pub.schema.json
Comment (frrist) [reply schemas/job/schemas/pub.schema.json:10]: https://github.com/bacalhau-project/bacalhau/pull/4074#discussion_r1640276157

File: schemas/job/schemas/strg-local.schema.json
Comment (frrist) [reply schemas/job/schemas/strg-local.schema.json:6]: I wasn't aware of this, will address. What does it mean for `ReadWrite` to be false? Does it mean the local storage cannot be read or written to? Similarly what does it mean for it to be true?

**Comment by aronchick** [comment]
No - for SURE we want a json schema of our job spec. In order to be a great platform, we need people externally to us to be submitting jobs. Offering a hosted job back that is the exact same one as we use internally is a very big positive.

**Comment by aronchick** [comment]
I understand about not wanting to cause big changes the week before we launch, but for better or worse, this looks like the right thing to me. @wdbaruni can you comment on what your reservation is here?

**Comment by frrist** [comment]
And just so its clear, this schema-based validation is only used by the code path of the `job validate` command and not anywhere else.

File: pkg/lib/schemaloader/embedded.go
Comment (frrist) [reply pkg/lib/schemaloader/embedded.go:48]: I suppose we don't need to care, you're right. I can drop this check.

File: schemas/job/schemas/strg-local.schema.json
Comment (wdbaruni) [reply schemas/job/schemas/strg-local.schema.json]: It means readonly access

Comment (frrist) [reply schemas/job/schemas/strg-local.schema.json]: In the future, do you think it would be worth changing the name of the field to `ReadOnly`, `Write`, or `Writeable`?

Comment (wdbaruni) [reply schemas/job/schemas/strg-local.schema.json]: As long as it is easy to default to `ReadOnly`, then maybe _in the future_

**Comment by wdbaruni** [comment]
> And just so its clear, this schema-based validation is only used by the code path of the `job validate` command and not anywhere else.

And that makes things even worse as this whole complexity is just for a command that is barely used and follows a different validation path than in the server side or `job run`, but I don't think the intention is to use this schema validation only in `job validate`

**Comment by wdbaruni** [comment]
> No - for SURE we want a json schema of our job spec. In order to be a great platform, we need people externally to us to be submitting jobs. Offering a hosted job back that is the exact same one as we use internally is a very big positive.
> I understand about not wanting to cause big changes the week before we launch, but for better or worse, this looks like the right thing to me. @wdbaruni can you comment on what your reservation is here?

This is a general statement Dave. We need to understand what problem we want to solve or improve, and what is the right time and way to deliver it. It seems from the answers above, the only value is external integrations, and this is not the right time to solve that or evaluate the right way to achieve it.

My reservation is that anything we introduce is a cost that we have to maintain, and I am not sure why we keep on introducing things without any immediate value or user demand, and adds these complexities to us:
1. Forking the validation path as it seems we can't do all validations against the schema, and will continue to maintain both paths.
2. As a developer, I'll need to maintain both paths, not clear where to encode new validations, and we will need to make sure the json schema is always in sync
3. We are not really improving the user experience as `validate` can pass while `run` can fail
4. We are restricting the engines that we support, and forcing the user to update the client if any engine is introduced, which goes against our goal of being agnostic to the engines at the compute side
5. We just introduced json schemas for the job spec, but for integration purposes we need them for all of our APIs. We can't keep adding and maintaining these schemas manually. We need to look for ways to automate their creation out of our golang types.

All of these complexities are worth owning and maintaining when they are worth solving and have demand for the values that they add, but that is not today. I would also like to explore auto-generation of json schemas from our structs instead of encoding them manually, but only when there is a demand for them.

We have already spent some time implementing CUE lang  schema validations before that didn't make it through for some reason, now we are suggesting json schema instead for no apparent reason and no evaluation of alternatives. These types of distractions is what is concerning me. We are less than a week away from 1.4, the features are not all done yet, and we shouldn't add changes last minute. In fact we should've already frozen 1.4 but we can't as we are behind.

**Comment by frrist** [comment]
> Didn't you add a similar experience in https://github.com/bacalhau-project/bacalhau/pull/4103 by providing a custom unmarshaller?

Yup, #4103 is the fall back plan for this work remaining in draft. Will open a new PR that uses that method instead.

**Pull request closed without merging**
